File: app/clojure/part1/src/immutability.clj
(ns immutability
  "第1章: 不変性とデータ変換

   関数型プログラミングにおける不変データ構造の基本から、
   データ変換パイプライン、副作用の分離までを扱います。")

;; =============================================================================
;; 1. 不変データ構造の基本
;; =============================================================================

(defn update-age
  "人物の年齢を更新する（元のデータは変更しない）"
  [person new-age]
  (assoc person :age new-age))

;; =============================================================================
;; 2. 構造共有（Structural Sharing）
;; =============================================================================

(defn add-member
  "チームに新しいメンバーを追加する"
  [team member]
  (update team :members conj member))

;; =============================================================================
;; 3. データ変換パイプライン
;; =============================================================================

(defn calculate-subtotal
  "各アイテムの小計を計算する"
  [item]
  (* (:price item) (:quantity item)))

(defn membership-discount
  "会員種別に応じた割引率を取得する"
  [membership]
  (case membership
    :gold 0.1
    :silver 0.05
    :bronze 0.02
    0))

(defn calculate-total
  "注文の合計金額を計算する"
  [order]
  (->> (:items order)
       (map calculate-subtotal)
       (reduce +)))

(defn apply-discount
  "割引後の金額を計算する"
  [order total]
  (let [discount-rate (-> order :customer :membership membership-discount)]
    (* total (- 1 discount-rate))))

(defn process-order
  "注文を処理し、割引後の合計金額を返す"
  [order]
  (let [total (calculate-total order)]
    (apply-discount order total)))

;; =============================================================================
;; 4. 副作用の分離
;; =============================================================================

(defn pure-calculate-tax
  "純粋関数：税額を計算する"
  [amount tax-rate]
  (* amount tax-rate))

(defn calculate-invoice
  "ビジネスロジック（純粋関数）：請求書を計算する"
  [items tax-rate]
  (let [subtotal (reduce + (map calculate-subtotal items))
        tax (pure-calculate-tax subtotal tax-rate)
        total (+ subtotal tax)]
    {:subtotal subtotal
     :tax tax
     :total total}))

;; 副作用を含む処理（分離）
(defn save-invoice!
  "データベースへの保存（副作用）"
  [invoice]
  (println "Saving invoice:" invoice)
  invoice)

(defn send-notification!
  "メール送信（副作用）"
  [invoice customer-email]
  (println "Sending notification to:" customer-email)
  invoice)

(defn process-and-save-invoice!
  "処理全体のオーケストレーション"
  [items tax-rate customer-email]
  (-> (calculate-invoice items tax-rate)
      save-invoice!
      (send-notification! customer-email)))

;; =============================================================================
;; 5. 永続的データ構造の活用：Undo/Redo の実装
;; =============================================================================

(defn create-history
  "履歴を保持するデータ構造を作成する"
  []
  {:current nil
   :past []
   :future []})

(defn push-state
  "新しい状態を履歴にプッシュする"
  [history new-state]
  (-> history
      (update :past conj (:current history))
      (assoc :current new-state)
      (assoc :future [])))

(defn undo
  "直前の状態に戻す"
  [history]
  (if (empty? (:past history))
    history
    (let [previous (peek (:past history))]
      (-> history
          (update :future conj (:current history))
          (assoc :current previous)
          (update :past pop)))))

(defn redo
  "やり直し操作"
  [history]
  (if (empty? (:future history))
    history
    (let [next-state (peek (:future history))]
      (-> history
          (update :past conj (:current history))
          (assoc :current next-state)
          (update :future pop)))))

;; =============================================================================
;; 6. トランスデューサーによる効率的な変換
;; =============================================================================

(def xf-process-items
  "複数の変換を合成したトランスデューサー"
  (comp
   (filter #(> (:quantity %) 0))
   (map #(assoc % :subtotal (calculate-subtotal %)))
   (filter #(> (:subtotal %) 100))))

(defn process-items-efficiently
  "トランスデューサーを使用した効率的な処理"
  [items]
  (into [] xf-process-items items))

File: app/haskell/part1/src/Immutability.hs
{-|
Module      : Immutability
Description : Chapter 1 - Immutability and Data Transformation
Copyright   : (c) k2works, 2025
License     : MIT

This module demonstrates immutability concepts in Haskell:
- Immutable data structures
- Structural sharing
- Data transformation pipelines
- Side effect separation
- History management (Undo/Redo)
-}
module Immutability
    ( -- * Basic Data Types
      Person(..)
    , updateAge
      -- * Team Management
    , Member(..)
    , Role(..)
    , Team(..)
    , addMember
      -- * Order Processing
    , Item(..)
    , Membership(..)
    , Customer(..)
    , Order(..)
    , calculateSubtotal
    , membershipDiscount
    , calculateTotal
    , applyDiscount
    , processOrder
      -- * Invoice Processing
    , Invoice(..)
    , calculateInvoice
      -- * History Management
    , History(..)
    , createHistory
    , pushState
    , undo
    , redo
    , currentState
    , canUndo
    , canRedo
    ) where

-- | Person record demonstrating basic immutability
data Person = Person
    { personName :: String
    , personAge  :: Int
    } deriving (Show, Eq)

-- | Update a person's age, returning a new Person (original unchanged)
updateAge :: Person -> Int -> Person
updateAge person newAge = person { personAge = newAge }

-- | Member roles in a team
data Role = Developer | Designer | Manager
    deriving (Show, Eq)

-- | Team member
data Member = Member
    { memberName :: String
    , memberRole :: Role
    } deriving (Show, Eq)

-- | Team with structural sharing
data Team = Team
    { teamName    :: String
    , teamMembers :: [Member]
    } deriving (Show, Eq)

-- | Add a member to a team (structural sharing - existing members are shared)
addMember :: Team -> Member -> Team
addMember team newMember = team { teamMembers = teamMembers team ++ [newMember] }

-- | Order item
data Item = Item
    { itemName     :: String
    , itemPrice    :: Int
    , itemQuantity :: Int
    } deriving (Show, Eq)

-- | Membership levels for customers
data Membership = Gold | Silver | Bronze | Regular
    deriving (Show, Eq)

-- | Customer information
data Customer = Customer
    { customerName       :: String
    , customerMembership :: Membership
    } deriving (Show, Eq)

-- | Order with items and customer
data Order = Order
    { orderItems    :: [Item]
    , orderCustomer :: Customer
    } deriving (Show, Eq)

-- | Calculate subtotal for a single item
calculateSubtotal :: Item -> Int
calculateSubtotal item = itemPrice item * itemQuantity item

-- | Get discount rate based on membership level
membershipDiscount :: Membership -> Double
membershipDiscount Gold   = 0.1
membershipDiscount Silver = 0.05
membershipDiscount Bronze = 0.02
membershipDiscount Regular = 0.0

-- | Calculate total for an order (before discount)
calculateTotal :: Order -> Int
calculateTotal order = sum $ map calculateSubtotal (orderItems order)

-- | Apply membership discount to total
applyDiscount :: Order -> Int -> Double
applyDiscount order total =
    let discountRate = membershipDiscount (customerMembership (orderCustomer order))
    in fromIntegral total * (1 - discountRate)

-- | Process order: calculate total with discount applied
processOrder :: Order -> Double
processOrder order =
    let total = calculateTotal order
    in applyDiscount order total

-- | Invoice data
data Invoice = Invoice
    { invoiceSubtotal :: Int
    , invoiceTax      :: Double
    , invoiceTotal    :: Double
    } deriving (Show, Eq)

-- | Calculate invoice from items and tax rate (pure function)
calculateInvoice :: [Item] -> Double -> Invoice
calculateInvoice items taxRate =
    let subtotal = sum $ map calculateSubtotal items
        tax = fromIntegral subtotal * taxRate
        total = fromIntegral subtotal + tax
    in Invoice subtotal tax total

-- | History state for undo/redo functionality
data History a = History
    { historyCurrent :: Maybe a
    , historyPast    :: [a]
    , historyFuture  :: [a]
    } deriving (Show, Eq)

-- | Create empty history
createHistory :: History a
createHistory = History Nothing [] []

-- | Push a new state onto history
pushState :: History a -> a -> History a
pushState history newState =
    History
        { historyCurrent = Just newState
        , historyPast = case historyCurrent history of
            Nothing -> historyPast history
            Just current -> current : historyPast history
        , historyFuture = []
        }

-- | Undo to previous state
undo :: History a -> History a
undo history = case historyPast history of
    [] -> history
    (prev:rest) -> History
        { historyCurrent = Just prev
        , historyPast = rest
        , historyFuture = case historyCurrent history of
            Nothing -> historyFuture history
            Just current -> current : historyFuture history
        }

-- | Redo to next state
redo :: History a -> History a
redo history = case historyFuture history of
    [] -> history
    (next:rest) -> History
        { historyCurrent = Just next
        , historyPast = case historyCurrent history of
            Nothing -> historyPast history
            Just current -> current : historyPast history
        , historyFuture = rest
        }

-- | Get current state from history
currentState :: History a -> Maybe a
currentState = historyCurrent

-- | Check if undo is possible
canUndo :: History a -> Bool
canUndo history = not $ null $ historyPast history

-- | Check if redo is possible
canRedo :: History a -> Bool
canRedo history = not $ null $ historyFuture history
